File: datasets/Python/Claude/Algorithm/hard4.py
from collections import deque

def solve_reverse_factorization():
    # Manual input - user enters values during runtime
    print("Enter N and K (space separated):")
    n, k = map(int, input().split())
    print(f"Enter {k} elements of array A (space separated):")
    a = list(map(int, input().split()))
    
    # Sort the array to ensure lexicographic ordering
    a.sort()
    
    # BFS to find shortest path
    queue = deque([(1, [1])])  # (current_value, path)
    visited = {1}
    
    while queue:
        current, path = queue.popleft()
        
        if current == n:
            print(*path)
            return
        
        # Try multiplying with each element in sorted order
        for multiplier in a:
            next_val = current * multiplier
            
            # Pruning: if next_val > n, skip
            if next_val > n:
                continue
                
            # If not visited, add to queue
            if next_val not in visited:
                visited.add(next_val)
                new_path = path + [next_val]
                queue.append((next_val, new_path))
    
    # If we reach here, no solution exists
    print(-1)

# Example usage with manual input
if __name__ == "__main__":
    print("=== Reverse Factorization Problem ===")
    print("Find the lexicographically smallest path with minimum states from 1 to N")
    print("Example: N=12, A=[2,3,4] should give output: 1 3 12")
    print()
    solve_reverse_factorization()

File: datasets/C++/Claude/Algorithm/hard4.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
#include <algorithm>
using namespace std;

void solve() {
    // Manual input - user enters values during runtime
    cout << "Enter N and K (space separated): ";
    int n, k;
    cin >> n >> k;
    
    cout << "Enter " << k << " elements of array A (space separated): ";
    vector<int> a(k);
    for (int i = 0; i < k; i++) {
        cin >> a[i];
    }
    
    // Sort to ensure lexicographic ordering
    sort(a.begin(), a.end());
    
    // BFS
    queue<pair<long long, vector<long long>>> q;
    unordered_set<long long> visited;
    
    q.push({1, {1}});
    visited.insert(1);
    
    while (!q.empty()) {
        auto [current, path] = q.front();
        q.pop();
        
        if (current == n) {
            for (int i = 0; i < path.size(); i++) {
                if (i > 0) cout << " ";
                cout << path[i];
            }
            cout << endl;
            return;
        }
        
        // Try each multiplier in sorted order
        for (int multiplier : a) {
            long long next_val = current * multiplier;
            
            // Pruning
            if (next_val > n) continue;
            
            if (visited.find(next_val) == visited.end()) {
                visited.insert(next_val);
                vector<long long> new_path = path;
                new_path.push_back(next_val);
                q.push({next_val, new_path});
            }
        }
    }
    
    cout << -1 << endl;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    cout << "=== Reverse Factorization Problem ===" << endl;
    cout << "Find the lexicographically smallest path with minimum states from 1 to N" << endl;
    cout << "Example: N=12, A=[2,3,4] should give output: 1 3 12" << endl;
    cout << endl;
    
    solve();
    
    return 0;
}